### ISSUE TITLE
Machine Learning Acceleration

### ISSUE THREAD

**Issue by raduweiss** [firecracker-microvm/firecracker#1179]
Doing hardware-accelerated inference in a serverless environment is compelling use case. 

However, adding straight up GPU passthrough means that microVM can't oversubscribe memory, and we need to add PCI emulation to Firecracker, which comes with a lot of extra complexity/attack surface.

The first step here will be to research the options and alternatives (e.g., GPU passthrough, or something else), and figure out the path forward.

Related issues: #849, #776.

**Comment by nlauchande** [comment]
I am very interested on this usecase.

**Comment by richardliaw** [comment]
+1, very interested in this use case. Any update on this? (I understand it's still in the research phase)

**Comment by zaharidichev** [comment]
@raduweiss  is this something that anyone is working on atm? Is it still on the roadmap?

**Comment by ananos** [comment]
Hi @zaharidichev,

we have some thoughts on this [1], shared them earlier this year in the slack wοrkspace [2], but a chat is still pending I'm afraid. We have a rough proof-of-concept implementation on firecracker, based on the design principles of [1], which exhibits negligible overhead for image inference (jetson-inference backend, using tensorRT, tested on an NVIDIA jetson nano & a generic x86_64 machine with an RTX 2060 SUPER & another machine with a T4). We should be able to open-source the whole stack pretty soon. Feel free to drop us a line if you're interested in our early PoC.

Essentially, the idea is that we abstract away the hardware-specific operations via a slim runtime library/system, that supports any kind of backend (ranging from a simple CUDA/OpenCL function to a TensorFlow operation/app). Combined with a simple virtio frontend/backend implementation we are able to forward operations from a guest to the host/monitor, which in turn executes the actual "acceleratable" function on the hardware accelerator.

Another option (if latency is not critical to you) could be to use rCUDA, which we plan to try but haven't had the time yet...  

BTW, @raduweiss we should plan to have that chat [2] at some point -- give us a shout when you are available!  

cheers,
Tassos

[1] https://blog.cloudkernels.net/posts/vaccel/
[2] https://firecracker-microvm.slack.com/archives/CDL3FUR8B/p1591093992140800

**Comment by raduweiss** [comment]
@ananos , yeah our bad, we totally dropped the ball here. Our apologies! I'll reply directly so we can talk.

**Comment by ananos** [comment]
Hi @zaharidichev, all

just wanted to share our blog post about our approach on the above: https://blog.cloudkernels.net/posts/vaccel_v2/

using nvidia-container-runtime & a docker image we've put together, you are able to run the jetson-inference image classification example from a Firecracker VM. You can find more info in the above post or @ https://vaccel.org. Of course, you can ping us, we will be more than happy to share how to try out vAccel on Firecracker.

cheers,
Tassos

**Comment by amrragab8080** [comment]
Any update on the GPU support in Firecracker?

**Comment by raduweiss** [comment]
We’ve been thinking about / experimenting in this space in the last months, and we'll keep at it this year, but there’s no ETA for this feature right now. For maximum utility in a serverless platform paradigm [a], a single GPU hardware resource needs to be safely used by multiple microVMs, without trading off the other capabilities that Firecracker users like (e.g., CPU/memory oversubscription, fast snapshot-restore, or high mutation rate of the host’s microVMs). This is a pretty complex problem, and we’re still exploring our options.
 
We’d be happy to hear of any use cases to so we can factor them in – feel free to update this thread, or share them directly on our Slack [b]!
 
[a] https://github.com/firecracker-microvm/firecracker/blob/master/CHARTER.md
[b] [firecracker-microvm Slack workspace link](https://join.slack.com/t/firecracker-microvm/shared_invite/enQtNDY2NTUwMzQ3MDE1LWIwMzA0OWFkMTZhMTlmMDZiMmFkYjMyODMxMGQ1ZjliMzJjNjJiNWRhNWNkOGEyNmUxNmRkMjZhYTc3MmVjZjM)

**Comment by pdames** [comment]
Any updates? My team is interested in running [Ray](https://github.com/ray-project/ray) on Firecracker, but the current lack of GPU support would erode the value of doing so.

**Comment by raduweiss** [comment]
> Any updates? My team is interested in running [Ray](https://github.com/ray-project/ray) on Firecracker, but the current lack of GPU support would erode the value of doing so.

Sorry for not getting back here sooner, we were still working through our options. We've settled on implementing plain PCIe GPU passthrough, which comes at the cost of requiring micoVMs to start with the their full memory mapped, will probably negate the advantages of using snapshot-restore, and requires the full GPU to be attached to a microVM - all things we wanted to see if we could improve upon, but we didn't find way that upholds all our tenets.

We will want to get broad feedback from the community here on how to actually present this as a feature (we'll start a discussion in the following weeks). Given the trade-offs above, we will consider building a separate Firecracker mode or Firecracker variant, or something along those lines.

**Comment by zvonkok** [comment]
@raduweiss I am leading the enablement of GPUs and other NV accelerators on Kata containers. I was trying to use the Slack Invite in the README.md but it is invalid. 

What would be the best way to get into the loop on the PCIe implementation in firecracker? I fixed and I'm currently fixing several other issues (BAR sizes, MDEV support, ...) in Kata's PCIe (QEMU) implementation. 

Would be nice if I could get hands-on with some pre-released artifacts to start testing on our side.

**Comment by DemiMarie** [comment]
@raduweiss: what would be needed for a “good” solution?  Could https://libvf.io be helpful?

**Comment by mmcclean-aws** [comment]
Any plans to support Inferentia and Trainium based instances ? They expose the accelerators via PCI to the OS but I see PCI support is not planned for firecracker

**Comment by kalyazin** [comment]
Hi @mmcclean-aws . Like discussed offline, an immediate obstacle for supporting Inferentia and Trainium instances is that they are virtualised (as opposed to bare metal), so Firecracker can't run on them, because AWS doesn't support nested virtualisation. Besides that, since Inf2 has 12 accelerators, and each accelerator can only be used in a single-tenant manner, the instance can carry up to 12 microVMs at the same time, which does not allow to extract oversubscription that is a key Firecracker's benefit. The only potential benefit (if/when bare metal Inf2* instances are available) could be shorter VM startup time if an instance needs to be partitioned dynamically.

**Comment by peterdelevoryas** [comment]
> The only potential benefit (if/when bare metal Inf2* instances are available) could be shorter VM startup time if an instance needs to be partitioned dynamically.

I wouldn't say this is the only benefit. I want to replace QEMU with something more efficient and modern. Rust is a huge draw. I don't care about oversubscription, that's not what interests me.

_A lot_ of people are working on accelerated computing now. We're all taking machines with 12+ accelerators and trying to virtualize them. I _need_ support for passthrough, I don't have any workloads that don't include accelerators.

**Comment by kalyazin** [comment]
Hi @peterdelevoryas . What is your specific motivation for moving off Qemu? Is that merely because of the Rust safety features? Firecracker was developed with CPU workloads in mind, and design decisions have been often driven by that (eg using MMIO virtio transport vs PCI). Being a live product, we may find it possible to reconsider those if sufficiently compelling reasons for doing so arise.

**Comment by peterdelevoryas** [comment]
> Hi @peterdelevoryas . What is your specific motivation for moving off Qemu? Is that merely because of the Rust safety features? Firecracker was developed with CPU workloads in mind, and design decisions have been often driven by that (eg using MMIO virtio transport vs PCI). Being a live product, we may find it possible to reconsider those if sufficiently compelling reasons for doing so arise.

I don’t have any super strong reasons to migrate off QEMU, I just like the idea of something stripped down and written in Rust, and the fact it’s completely open source, free, and run in production for real aws workloads. I just don’t want to live with QEMU forever, even if just for the fact that I don’t enjoy mailing list development.

**Comment by DemiMarie** [comment]
A few comments:

- https://libvf.io provides support for GPU virtualization.  However, multi-tenant GPU virtualization requires trusting the proprietary vendor hardware & firmware to do its job.  @raduweiss: Does Amazon consider this sufficent protection?
- With PCI passthrough, Amazon can avoid most of the security risks by using a custom board design where the GPU’s SPI flash is write-protected by hardware the GPU has no control of.  Passthrough with stock hardware is much riskier.
- Memory oversubscription is possible by emulating a nested IOMMU.  I’m not sure if Firecracker’s developers are interested in doing so given the performance penalties.

**Comment by mmcclean-aws** [comment]
Thanks. Does that mean that PCI passthrough should work for alternative devices (e.g. Trainium and Inferentia) that expose themselves in `/dev` ?

**Comment by DemiMarie** [comment]
It should _work_ for almost any PCI device.  Whether it is _secure_ is another matter.  _That_ depends entirely on choosing a safe device and your ability to prevent early boot DMA attacks and unintended persistence via e.g. on-device flash storage.

**Comment by jayavanth** [comment]
Any updates on GPU support in your roadmap for 2023/2024?

**Comment by xmarcalx** [comment]
Hi @jayavanth ,

Thanks for your question.
No we are not planning any GPU support in Firecracker at the moment. 
Once we will consider again this task we will add in our GitHub roadmap, which we are in the following weeks we are going to update and bring up to speed soon.

**Comment by fighterhit** [comment]
> Hi @jayavanth ,
> 
> Thanks for your question. No we are not planning any GPU support in Firecracker at the moment. Once we will consider again this task we will add in our GitHub roadmap, which we are in the following weeks we are going to update and bring up to speed soon.

Hi @xmarcalx , in the current era of rapid AI development, GPU support is very important. I hope the team can seriously consider this feature. Thanks!

**Comment by Talador12** [comment]
At some point, it was decided that PCI passthrough would be acceptable to the tenets of firecracker
https://github.com/firecracker-microvm/firecracker/issues/1179#issuecomment-872763765

This could actually work well with Nvidia's MIG slice paradigm. You would still have to passthrough a physical slice of the GPU, but you could guarantee that slice to be isolated within that allocation
https://docs.nvidia.com/datacenter/tesla/mig-user-guide/index.html#virtualization

**Comment by DemiMarie** [comment]
The best way I know to scalably oversubscribe a GPU is virtio-GPU native contexts, which is what Qubes OS will be using for GPU virtualization in the future.  Native contexts expose a subset of the host kernel driver API, which means that the number of clients is limited by available memory rather than by GPU hardware or firmware.  Native contexts neither use nor rely on hardware or firmware GPU virtualization suppot, so they work with any GPU that has the needed support in Mesa and virglrenderer.  Proprietary drivers (such as Nvidia) are not supported.

On AMD GPUs, process isolation mode is required for security.

**Comment by xmarcalx** [comment]
Dear @Talador12 et all,

Sorry for the late reply, and thank you very much for your interest in Firecracker and specifically into this feature.
GPUs support is a long due feature request coming from the community. We are always interested into learn more about potential use-case which justify GPU support for Firecracker workloads to make sure that it is the right thing for the product and the application.
As per previous message, we were not planning to support GPUs in Firecracker but we are more than happy to support the community in this effort, particularly around:

1. building a plan for the community to develop a first PoC which include PCIe and GPU support into Firecracker microVMs
2. analyze at each milestone of the project performance KPI and security stance of Firecracker microVMs
3. and work together to define a plan on how to deliver these milestone into Firecracker releases.

If you are onboard, we should set up an introductory call within the next couple of weeks with you and any other interested users to understand the initial requirements and usecase and layout the next steps.
We will announce the schedule of this meeting in our slack channel and in our GitHub discussions section.
I look forward to hearing from you soon.

Kind Regards,
Marco

**Comment by Talador12** [comment]
@xmarcalx Hey Marco,
For the meeting, September 30th / October 9th work for us. Which would you prefer?
Also - we should include interested users in this thread/slack as well.

**Comment by xmarcalx** [comment]
Hi @Talador12 ,

sure, let's organize for the 9th October. 
From which timezone are the participants?
Such that we can try to organize a time friendly meeting for at least the majority of people. 

For any interested folk in the matter, keep posted i will post a link to the meeting who will be free for everyone to join  in this issue and in the slack thread in our community https://firecracker-microvm.slack.com/archives/CDL3FUR8B/p1724175189998039

**Comment by xmarcalx** [comment]
Hi everyone,

we set up the meeting for the 9th October, from 18:00 BST to 19:00 BST to:

1.  dive deep on motivation, technical requirements of this request 
2. discuss the next steps we can plan with the community to move forward with this ask.

The meeting details are following:

You have been invited to an online meeting, powered by Amazon Chime.
Click to join the meeting: https://chime.aws/7995191427
Meeting ID: 7995 19 1427
A headset is recommended or you may use your computer’s microphone and speakers.
Call in using your phone:
United States Toll-Free (1): [+1 855-552-4463](tel:+1855-552-4463)
Meeting ID: 7995 19 1427
One-click Mobile Dial-in (United States Toll-Free (1)): [+1 855-552-4463,,,7995191427#](tel:+1855-552-4463,,,7995191427)
United Kingdom Toll-Free (1): [+44 800 085 5175](tel:+44800%20085%205175)
International: https://chime.aws/dialinnumbers/
Dial-in attendees must enter *7 to mute or unmute themselves.
To connect from an in-room video system, use one of the following Amazon Chime bridges:
SIP video system: 7995191427@meet.chime.in or meet.chime.in
H.323 system: 13.248.147.139 or 76.223.18.152
If prompted enter the Meeting PIN: 7995191427#
Download Amazon Chime at https://aws.amazon.com/chime/download
For information about creating an Amazon Chime account, see https://aws.amazon.com/chime/getting-started

See you there!

**Comment by Talador12** [comment]
We will see you there! Looking forward to it :)

**Comment by xmarcalx** [comment]
we decided to track the note of the meeting in this discussion #4845. The discussion contains also the link to the meeting we will use to sync every 4 weeks.

**Comment by banrovegrie** [comment]
Hi everyone, any further updates on GPU and PCIe support?

**Comment by DemiMarie** [comment]
I don’t think so.  PCI passthrough loses some of the features that makes Firecracker so attractive.  Also, depending on threat model one might need to perform a full GPU power cycle before reassigning the GPU to a different tenant.  It’s worth noting that Amazon’s own SageMaker AI Serverless does not support GPU acceleration, only CPU inference.

Cloudflare Workers AI took a different approach, which is to only support a predefined set of trusted models.  This allows them to serve multiple tenants without having to assign GPUs to their VMs.

Unfortunately, the things that make GPUs fast make this hard.  GPU page faults are far slower than their CPU counterparts.  GPUs have a very large amount of non-virtualized on-die state, which makes context switches extremely expensive.  Furthermore, this state cannot be partitioned between threads, so generally one can only execute one customer workload on the GPU at a time.

I don’t see the situation changing any time soon.  If cloud providers trusted SR-IOV to isolate tenants from each other, things _might_ be different.  However, my understanding is they currently do not, and presumably have a good reason not to.  Virtio-GPU native contexts exposes some of the kernel driver attack surface and cannot prevent denial of service, so it isn’t a good option either.  On at least AMD, I don’t think all GPU hangs are guaranteed to be recoverable, at least without crashing other workloads.  See LeftoverLocals and Whispering Pixels for attacks that can happen if kernel driver isolation is used.

The best approach I can think of for hardware-accelerated serverless inference is to harden an ML inference server sufficiently that it can safely execute untrusted ML models.  There are model formats that should not allow executing malicious code, such as SafeTensors.

**Comment by xmarcalx** [comment]
Hi @banrovegrie,

Thanks for the message. 

Firecracker added PCIe support since our latest release v1.13.x so if you need that functionality is already out there.

Regarding GPU support, we are keeping monitoring the space. We didn't really progress and we are still researching around a ways to support GPUs while maintaining our security promises to our customers, with performance profiles which are still valid to target our ephemeral workloads and somehow being economical viable. We do not have date yet for official GPU support in Firecracker.

As @DemiMarie said, there are a lot of challenges in doing the right thing in the space with the technology available.

**Comment by sodabrew** [comment]
@xmarcalx to clarify my understanding, if you could help validating these examples:

Right now one server equipped with 8 GPU cards can assign each card as a PCIe passthrough device to a single Firecracker VM. In this example, one server, 8 VMs, all good.

What's not supported is preempting one of those VMs and switching to a different VM assigned to the same device because the device is very stateful and the size of the state is very large (e.g. many gigabytes of a vector dataset). In this example, one server, 16 VMs, 2 VMs per GPU, nope no good.

What's also not supposed is GPUs that have partitioning support, such as https://docs.nvidia.com/vgpu/deployment/azure-stack-hci/latest/gpup-workflow.html because the relevant drivers are in the Windows ecosystem. Right?

**Comment by DemiMarie** [comment]
If you need resource overcommit, virtio-GPU native contexts might be a good option.  The attack surface they expose is the GPU firmware and a subset of the host GPU driver.  Most vulnerabilities in the kernel driver aren’t exploitable via this route, as host userspace validates ioctl parameters and always makes calls from a single thread.  For instance, it will never close a file descriptor on one thread while an ioctl on it in another thread is in progress.  On AMD GPUs, you will need to enable process isolation.  This also helps mitigate the impact of a GPU hang.

However, native contexts currently only support graphics APIs, and only on AMD Qualcomm, and Apple GPUs.  Support for other GPUs has not started or is a work in progress.  Help with this would be greatly appreciated, though.

Denial of service is only a serious problem if you _expect_ to be running hostile workloads.  Otherwise, a workload that repeatedly crashes the GPU is only going to get itself shut down, which is the opposite of what malware wants 🙂.  However, serverless _does_ expect to be running potentially hostile workloads.  That’s a much harder situation.

For what it is worth, both Qubes OS and Spectrum will be using virtio-GPU native contexts or a similar technology, though in at least Qubes OS it will be off by default.  Both very much care about security, but the goal is to protect compromise of individual workloads from escalating to full system compromise.  It’s not to run arbitrary code of unknown origin.  This means that denial of service is not a serious problem.  Also, there is a human at the console who can take action if there is a problem.

I made a talk at Xen Project Summit 2024 about this.  I recommend watching it if you are interested in this problem.
